File: tithe_budget/Core/Speech/SpeechParser.swift
import Foundation

/// 語音辨識結果
struct ParsedResult {
    let amount: Double?
    let category: Category?
    let rawText: String

    var hasAmount: Bool { amount != nil && amount! > 0 }
    var hasCategory: Bool { category != nil }
}

/// 從語音辨識文字中提取金額與分類
final class SpeechParser {

    private static let categoryKeywords: [(String, Category)] = [
        ("十一", .tithe), ("孝親", .filial), ("交際", .social),
        ("住", .housing), ("還款", .debt), ("食行", .foodTransport),
        ("彈性", .flexible),
    ]

    static func parse(_ text: String) -> ParsedResult {
        var result = ParsedResult(amount: nil, category: nil, rawText: text)
        guard !text.trimmingCharacters(in: .whitespaces).isEmpty else { return result }

        var remaining = text

        // Step 1: 匹配分類關鍵字
        for (keyword, category) in categoryKeywords {
            if text.contains(keyword) {
                result = ParsedResult(amount: nil, category: category, rawText: text)
                remaining = text.replacingOccurrences(of: keyword, with: "").trimmingCharacters(in: .whitespaces)
                break
            }
        }

        // Step 2: 從剩餘文字提取金額
        if !remaining.isEmpty {
            result = ParsedResult(amount: AmountParser.parse(remaining), category: result.category, rawText: text)
        } else if result.category == nil {
            // 無分類時才嘗試從原始文字提取金額
            result = ParsedResult(amount: AmountParser.parse(text), category: nil, rawText: text)
        }

        return result
    }
}

File: faithful_steward/services/speech_parser.py
"""SpeechParser — 從語音辨識文字中提取金額與分類。

對應 iOS SpeechParser（Core/Speech/SpeechParser.swift）。
"""

from typing import Optional
from ..models import Category, ParsedResult
from .amount_parser import AmountParser


class SpeechParser:
    """
    從 SFSpeechRecognizer 回傳文字中提取金額 + 分類。
    策略：先匹配 Category 關鍵字，再從剩餘文字提取金額。
    """

    # 分類關鍵字映射（簡短名稱 → Category）
    _CATEGORY_KEYWORDS: dict[str, Category] = {
        cat.short_name: cat for cat in Category
    }

    @classmethod
    def parse(cls, text: str) -> ParsedResult:
        """
        解析語音辨識文字。

        Args:
            text: SFSpeechRecognizer 回傳的辨識文字

        Returns:
            ParsedResult: 包含 amount (float|None) 與 category (Category|None)

        Examples:
            "食行兩百五" → ParsedResult(amount=250, category=FOOD_TRANSPORT)
            "三百塊" → ParsedResult(amount=300, category=None)
            "十一" → ParsedResult(amount=None, category=TITHE)
        """
        result = ParsedResult(raw_text=text)

        if not text or not text.strip():
            return result

        text = text.strip()

        # Step 1: 匹配分類關鍵字
        remaining = text
        for keyword, category in cls._CATEGORY_KEYWORDS.items():
            if keyword in text:
                result.category = category
                remaining = text.replace(keyword, "", 1).strip()
                break

        # Step 2: 從剩餘文字提取金額
        if remaining:
            amount = AmountParser.parse(remaining)
            if amount is not None:
                result.amount = amount
        elif not result.has_category:
            # 只有當沒有匹配到分類時，才從原始文字提取金額
            # 避免分類關鍵字中包含數字（如「十一」）被誤解析為金額
            amount = AmountParser.parse(text)
            if amount is not None:
                result.amount = amount

        return result

    @classmethod
    def match_category(cls, text: str) -> Category | None:
        """僅匹配分類關鍵字（不回傳金額）"""
        for keyword, category in cls._CATEGORY_KEYWORDS.items():
            if keyword in text:
                return category
        return None
